File: examples/python/verify_signatures.py
"""Verify the signatures and digests of an RPM package."""

import argparse
import sys

from rpm_rs import Package, Verifier

parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument("rpm", help="Path to the RPM file")
parser.add_argument(
    "-k",
    "--key",
    action="append",
    default=[],
    help="Path to a public key file (ASCII-armored)",
)
args = parser.parse_args()

pkg = Package.open(args.rpm)

verifier = Verifier()
for key_file in args.key:
    verifier.load_from_asc_file(key_file)
has_key = bool(args.key)

report = pkg.check_signatures(verifier)
digests = report.digests

print("Digest verification:")
print(f"  Header SHA1      : {digests.header_sha1}")
print(f"  Header SHA256    : {digests.header_sha256}")
print(f"  Header SHA3-256  : {digests.header_sha3_256}")
print(f"  Payload SHA256   : {digests.payload_sha256}")
print(f"  Payload SHA512   : {digests.payload_sha512}")
print(f"  Payload SHA3-256 : {digests.payload_sha3_256}")

if not report.signatures:
    print("\nNo signatures found.")
else:
    print("\nSignatures:")
    for sig in report.signatures:
        info = sig.info
        algo = info.algorithm or "unknown"
        hash_algo = info.hash_algorithm or "unknown"
        fp = f" [{info.fingerprint}]" if info.fingerprint else ""
        status = "OK" if sig.is_verified() else f"FAILED: {sig.error}"
        print(f"  {algo} / {hash_algo}{fp} — {status}")

ok = report.is_ok() if has_key else digests.is_ok()
if ok:
    print("\nResult: OK")
else:
    print("\nResult: FAILED")
    sys.exit(1)

File: examples/verify-signatures.rs
use clap::Parser;
use rpm::{Package, signature::pgp::Verifier};
use std::path::PathBuf;

/// Verify the signatures and digests of an RPM package.
#[derive(Parser)]
#[command(version)]
struct Args {
    /// Path to the RPM file
    rpm: PathBuf,

    /// Path to a public key file (ASCII-armored)
    #[arg(short, long)]
    key: Option<PathBuf>,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    let pkg = Package::open(&args.rpm)?;

    let verifier = match &args.key {
        Some(path) => Verifier::from_asc_file(path)?,
        None => Verifier::new(),
    };

    let report = pkg.check_signatures(verifier)?;

    println!("Digest verification:");
    println!("  Header SHA1      : {:?}", report.digests.header_sha1);
    println!("  Header SHA256    : {:?}", report.digests.header_sha256);
    println!("  Header SHA3-256  : {:?}", report.digests.header_sha3_256);
    println!("  Payload SHA256   : {:?}", report.digests.payload_sha256);
    println!("  Payload SHA512   : {:?}", report.digests.payload_sha512);
    println!("  Payload SHA3-256 : {:?}", report.digests.payload_sha3_256);

    if report.signatures.is_empty() {
        println!("\nNo signatures found.");
    } else {
        println!("\nSignatures:");
        for sig in &report.signatures {
            let info = &sig.info;
            let algo = info
                .algorithm()
                .map_or("unknown".to_string(), |a| format!("{a:?}"));
            let hash = info
                .hash_algorithm()
                .map_or("unknown".to_string(), |h| format!("{h:?}"));
            print!("  {algo} / {hash}");
            if let Some(fp) = info.fingerprint() {
                print!(" [{fp}]");
            }
            if sig.is_verified() {
                println!(" — OK");
            } else if let Some(err) = &sig.error {
                println!(" — FAILED: {err}");
            }
        }
    }

    let ok = if args.key.is_some() {
        report.is_ok()
    } else {
        report.digests.is_ok()
    };

    if ok {
        println!("\nResult: OK");
    } else {
        println!("\nResult: FAILED");
        std::process::exit(1);
    }

    Ok(())
}
